### ISSUE TITLE
`int32` overflow in `embedding_bag(mode="max")` backward pass

### ISSUE THREAD

**Issue by MarcBerliner** [pytorch/pytorch#188467]
### 🐛 Describe the bug

Hi, I found an `int32` overflow issue with the backward pass of `embedding_bag(mode="max")`. In the MWE below, the `int64` case performs the correct forward/backward pass, but the `int32` case overflows on the backward pass

MWE (needs >16 GB VRAM):
```python
import torch
import torch.nn.functional as F

DEVICE = "cuda"

dim = 2 ** 16
r = 2 ** 15

def grad_at_r(idx_dtype):
    w = torch.zeros(r + 1, dim, device=DEVICE, requires_grad=True)
    
    # make the final row contain the matrix's max value
    with torch.no_grad():
        w[r, :] = 1.0
    
    idx = torch.tensor([0, r], device=DEVICE, dtype=idx_dtype)
    off = torch.tensor([0], device=DEVICE, dtype=idx_dtype)
    F.embedding_bag(idx, w, off, mode="max").sum().backward()
    return w.grad[r].clone()

grad_at_r_exact = torch.ones(dim, device=DEVICE)

grad_at_r_int64 = grad_at_r(torch.int64) # grad_at_r_int64 = ones(dim)
torch.testing.assert_close(grad_at_r_exact, grad_at_r_int64) # pass

grad_at_r_int32 = grad_at_r(torch.int32) # grad_at_r_int32 = zeros(dim)
torch.testing.assert_close(grad_at_r_exact, grad_at_r_int32) # fails, int32 drops gradient on the final row
```

### Versions

```
PyTorch version: 2.11.0+cu128
Is debug build: False
CUDA used to build PyTorch: 12.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.2 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04.1) 13.3.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.39

Python version: 3.12.3 (main, Mar 23 2026, 19:04:32) [GCC 13.3.0] (64-bit runtime)
Python platform: Linux-5.15.167.4-microsoft-standard-WSL2-x86_64-with-glibc2.39
Is CUDA available: True
CUDA runtime version: 12.8
CUDA_MODULE_LOADING set to: 
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 5090 (32607 MiB, sm_120)
Nvidia driver version: 591.86
cuDNN version: 9.19.0 (91900)
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        48 bits physical, 48 bits virtual
Byte Order:                           Little Endian
CPU(s):                               32
On-line CPU(s) list:                  0-31
Vendor ID:                            AuthenticAMD
Model name:                           AMD Ryzen 9 9950X3D 16-Core Processor
CPU family:                           26
Model:                                68
Thread(s) per core:                   2
Core(s) per socket:                   16
Socket(s):                            1
Stepping:                             0
BogoMIPS:                             8599.84
Virtualization:                       AMD-V
Hypervisor vendor:                    Microsoft
Virtualization type:                  full
L1d cache:                            768 KiB (16 instances)
L1i cache:                            512 KiB (16 instances)
L2 cache:                             16 MiB (16 instances)
L3 cache:                             96 MiB (1 instance)

Versions of relevant libraries:
[pip3] numpy==2.5.0
[pip3] torch==2.11.0+cu128
[pip3] triton==3.6.0
[conda] Could not collect
```

**Comment by MarcBerliner** [comment]
I have a fix for this issue on a working branch, where I found that a `static_cast<index_t>(...)` in the embedding CUDA code causes the overflow. I'm happy to make a PR including the fix and a large tensor regression test if this is labeled actionable

**Comment by NikhilDendeti** [comment]
I can reproduce this. Given dim=2**16 and r=2**15, dim * r == 2**31, which is exactly the int32 overflow boundary — this points to an index/offset computation in the embedding_bag "max" mode backward CUDA kernel doing row * dim (or similar) in 32-bit arithmetic when index_t is int32, instead of widening to 64-bit before the multiply. I'll dig into EmbeddingBag.cu's backward-max kernel and confirm, then send a fix (cast to index_t/int64_t before the multiply, or use static_cast<int64_t> on the offending operand) plus a regression test based on this repro. Assigning myself unless someone's already on it.
